internal/proxy/proxy.go

```diff
@@ -88,6 +96,25 @@ type Config struct {
 	Dialer cloudsql.Dialer
 }
 
+// DialOptions interprets appropriate dial options for a particular instance
+// configuration
+func (c *Config) DialOptions(i *InstanceConnConfig) []cloudsqlconn.DialOption {
```

Comment:
Following the dumb struct smart procedure idiom, I put DialOptions on `Config` instead of `InstanceConnConfig`.  I like it better here because two less tight couplings than what @kurtisvg proposed: 

- `InstanceConnConfig` doesn't need a pointer field to `Config`
-  the code doesn't need to set the a pointer to the correct parent `Config`